server/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -261,65 +261,58 @@ private void assertMappingVersion(
-    }
-
-    public DocumentMapper merge(String type, CompressedXContent mappingSource, MergeReason reason) {
-        return internalMerge(type, mappingSource, reason);
-    }
-
-    private synchronized DocumentMapper internalMerge(IndexMetadata indexMetadata, MergeReason reason) {
+    public DocumentMapper merge(IndexMetadata indexMetadata, MergeReason reason) {
         assert reason != MergeReason.MAPPING_UPDATE_PREFLIGHT;
         MappingMetadata mappingMetadata = indexMetadata.mapping();
         if (mappingMetadata != null) {
-            return internalMerge(mappingMetadata.type(), mappingMetadata.source(), reason);
+            return mergeAndApplyMappings(mappingMetadata.type(), mappingMetadata.source(), reason);
         }
         return null;
     }
 
-    private synchronized DocumentMapper internalMerge(String type, CompressedXContent mappings, MergeReason reason) {
-
-        DocumentMapper documentMapper;
-
-        try {
-            documentMapper = parse(type, mappings);
-        } catch (Exception e) {
-            throw new MapperParsingException("Failed to parse mapping: {}", e, e.getMessage());
-        }
-
-        return internalMerge(documentMapper, reason);
+    public DocumentMapper merge(String type, CompressedXContent mappingSource, MergeReason reason) {
+        return mergeAndApplyMappings(type, mappingSource, reason);
     }
 
-    private synchronized DocumentMapper internalMerge(DocumentMapper mapper, MergeReason reason) {
-
-        assert mapper != null;
-
-        // compute the merged DocumentMapper
-        DocumentMapper oldMapper = this.mapper;
-        DocumentMapper newMapper;
-        if (oldMapper != null) {
-            newMapper = oldMapper.merge(mapper.mapping(), reason);
-        } else {
-            newMapper = mapper;
-        }
-        newMapper.root().fixRedundantIncludes();
-        newMapper.validate(indexSettings, reason != MergeReason.MAPPING_RECOVERY);
-
+    private synchronized DocumentMapper mergeAndApplyMappings(String mappingType, CompressedXContent mappingSource, MergeReason reason) {
+        Mapping mapping = mergeMappings(mappingType, mappingSource, reason);
+        DocumentMapper newMapper = newDocumentMapper(mapping, reason);
         if (reason == MergeReason.MAPPING_UPDATE_PREFLIGHT) {
             return newMapper;
         }
-
-        // commit the change
         this.mapper = newMapper;
         assert assertSerialization(newMapper);
+        return newMapper;
+    }
 
+    private DocumentMapper newDocumentMapper(Mapping mapping, MergeReason reason) {
+        DocumentMapper newMapper = new DocumentMapper(indexSettings, indexAnalyzers, documentParser, mapping);
+        newMapper.root().fixRedundantIncludes();
+        newMapper.validate(indexSettings, reason != MergeReason.MAPPING_RECOVERY);
         return newMapper;
     }
 
+    private synchronized Mapping mergeMappings(String mappingType, CompressedXContent mappingSource, MergeReason reason) {
```

Comment:
I don't think this needs to be `synchronized`, given that it's only called from `mergeAndApplyMappings()` which is already `synchronized`?

Reply:
that may totally be a leftover as the previous internalMerge were all synchronized